File: legacy/python/manul_engine/imports.py
# manul_engine/imports.py
"""
Hunt file import resolver and package manager.

Resolves @import: directives, manages hunt_libs/ packages,
and handles huntlib.json manifests.

Resolution order for import source paths:
  1. Relative to the importing .hunt file's directory
  2. Relative to CWD
  3. hunt_libs/<name>/  (looks for main.hunt or huntlib.json entry)
  4. ~/.manul/hunt_libs/<name>/  (installed scoped packages)
"""

from __future__ import annotations

import json
import os
import re
from typing import NamedTuple

from .exceptions import HuntImportError

_MAX_IMPORT_DEPTH = 10


class ImportDirective(NamedTuple):
    """Parsed @import: header directive."""

    block_names: list[str]  # ["Login", "Logout"] or ["*"]
    source: str  # "lib/auth_flows.hunt" or "@manul/saucedemo-flows"
    aliases: dict[str, str]  # {"Login": "AuthLogin"} when 'as' syntax used


# ── @import: header regex ────────────────────────────────────────────────────
# Matches:
#   @import: Login from lib/auth.hunt
#   @import: Login, Logout from lib/auth.hunt
#   @import: Login as AuthLogin from lib/auth.hunt
#   @import: * from lib/auth.hunt
_RE_IMPORT = re.compile(
    r"^@import:\s+(.+?)\s+from\s+(\S+)\s*$",
    re.IGNORECASE,
)


def parse_import_directive(line: str) -> ImportDirective | None:
    """Parse a single @import: line into an ImportDirective, or None."""
    m = _RE_IMPORT.match(line.strip())
    if m is None:
        return None

    names_part = m.group(1).strip()
    source = m.group(2).strip()

    if names_part == "*":
        return ImportDirective(block_names=["*"], source=source, aliases={})

    # Parse comma-separated block names with optional 'as Alias'
    block_names: list[str] = []
    aliases: dict[str, str] = {}

    for token in names_part.split(","):
        token = token.strip()
        if not token:
            continue
        # Check for 'as' alias: "Login as AuthLogin"
        as_match = re.match(r"^(.+?)\s+as\s+(\S+)$", token, re.IGNORECASE)
        if as_match:
            original = as_match.group(1).strip()
            alias = as_match.group(2).strip()
            block_names.append(original)
            aliases[original] = alias
        else:
            block_names.append(token)

    return ImportDirective(block_names=block_names, source=source, aliases=aliases)


def resolve_source_path(
    source: str,
    hunt_dir: str,
    cwd: str | None = None,
) -> str:
    """Resolve an import source path to an absolute file path.

    Resolution order:
      1. Relative to hunt_dir (the importing file's directory)
      2. Relative to CWD
      3. hunt_libs/<name>/main.hunt  (CWD-based)
      4. hunt_libs/<name>/  + huntlib.json entry field
      5. ~/.manul/hunt_libs/<name>/  (global install)

    Raises HuntImportError if no candidate exists.
    """
    cwd = cwd or os.getcwd()

    # If it looks like a file path (contains / or \ or ends with .hunt)
    # but NOT a scoped package like "@scope/pkg"
    is_file_path = ("/" in source or "\\" in source or source.endswith(".hunt")) and not source.startswith("@")
    if is_file_path:
        candidates = [
            os.path.normpath(os.path.join(hunt_dir, source)),
            os.path.normpath(os.path.join(cwd, source)),
        ]
        for c in candidates:
            if os.path.isfile(c):
                return os.path.abspath(c)
        tried = ", ".join(candidates)
        raise HuntImportError(f"Import source file not found: '{source}' (tried: {tried})")

    # Package-style import: "@manul/saucedemo-flows" or "auth_flows"
    # Normalize scoped names: @manul/foo → @manul/foo
    package_dirs = [
        os.path.join(cwd, "hunt_libs", source),
        os.path.join(os.path.expanduser("~"), ".manul", "hunt_libs", source),
    ]

    for pkg_dir in package_dirs:
        if not os.path.isdir(pkg_dir):
            continue
        # Check huntlib.json for entry field
        manifest_path = os.path.join(pkg_dir, "huntlib.json")
        if os.path.isfile(manifest_path):
            manifest = parse_huntlib_json(manifest_path)
            entry = manifest.get("entry", "main.hunt")
            entry_path = os.path.join(pkg_dir, entry)
            if os.path.isfile(entry_path):
                return os.path.abspath(entry_path)
        # Fallback: main.hunt
        main_path = os.path.join(pkg_dir, "main.hunt")
        if os.path.isfile(main_path):
            return os.path.abspath(main_path)

    tried = ", ".join(package_dirs)
    raise HuntImportError(f"Import package not found: '{source}' (tried: {tried})")


def parse_huntlib_json(path: str) -> dict:
    """Parse and validate a huntlib.json manifest file."""
    with open(path, encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise HuntImportError(f"huntlib.json must be a JSON object: {path}")
    return data


def _extract_exported_blocks(
    hunt_path: str,
    seen_files: set[str] | None = None,
) -> tuple[list[str], dict[str, list[str]], dict[str, str]]:
    """Parse a .hunt file and extract its exported STEP blocks.

    Returns (export_names, blocks_dict, parsed_vars) where:
      - export_names: list of @export: names (empty = nothing exported)
      - blocks_dict: {"BlockName": [action_lines...]} for all STEP blocks
      - parsed_vars: @var: declarations from the file

    Performs cycle detection via *seen_files*.
    """
    abs_path = os.path.abspath(hunt_path)
    if seen_files is not None and abs_path in seen_files:
        chain = " → ".join(sorted(seen_files)) + f" → {abs_path}"
        raise HuntImportError(
            f"Circular import detected: {chain}. Break the cycle by removing one of the @import: directives."
        )

    exports: list[str] = []
    parsed_vars: dict[str, str] = {}
    blocks: dict[str, list[str]] = {}
    current_block_name: str | None = None
    current_actions: list[str] = []
    in_setup = False
    in_teardown = False

    _re_step = re.compile(r"^\s*(?:\d+\.\s*)?STEP\s*\d*\s*:\s*(.*)", re.IGNORECASE)
    _re_setup = re.compile(r"^\s*\[SETUP\]", re.IGNORECASE)
    _re_end_setup = re.compile(r"^\s*\[END\s+SETUP\]", re.IGNORECASE)
    _re_teardown = re.compile(r"^\s*\[TEARDOWN\]", re.IGNORECASE)
    _re_end_teardown = re.compile(r"^\s*\[END\s+TEARDOWN\]", re.IGNORECASE)

    with open(hunt_path, encoding="utf-8") as fh:
        for line in fh:
            stripped = line.strip()

            # Skip hook blocks
            if _re_setup.match(stripped):
                in_setup = True
                continue
            if _re_end_setup.match(stripped):
                in_setup = False
                continue
            if _re_teardown.match(stripped):
                in_teardown = True
                continue
            if _re_end_teardown.match(stripped):
                in_teardown = False
                continue
            if in_setup or in_teardown:
                continue

            # Headers
            if stripped.startswith("@export:"):
                export_part = stripped.split(":", 1)[1].strip()
                if export_part:
                    for name in export_part.split(","):
                        cleaned = name.strip()
                        if cleaned:
                            exports.append(cleaned)
                continue
            if stripped.startswith("@var:"):
                var_part = stripped[5:].strip()
                m = re.match(r"\{?([^}=\s]+)\}?\s*=\s*(.*)", var_part)
                if m:
                    parsed_vars[m.group(1).strip()] = m.group(2).strip()
                continue
            if stripped.startswith("@"):
                continue
            if stripped.startswith("#") or not stripped:
                continue
            if stripped.upper().startswith("DONE"):
                continue

            # STEP marker
            step_m = _re_step.match(stripped)
            if step_m:
                # Save previous block
                if current_block_name is not None:
                    blocks[current_block_name] = current_actions
                current_block_name = step_m.group(1).strip()
                current_actions = []
                continue

            # Action line inside a block
            if current_block_name is not None:
                current_actions.append(stripped)

    # Save last block
    if current_block_name is not None:
        blocks[current_block_name] = current_actions

    return exports, blocks, parsed_vars


def resolve_imports(
    directives: list[ImportDirective],
    hunt_dir: str,
    hunt_file: str,
    cwd: str | None = None,
    seen_files: set[str] | None = None,
    _depth: int = 0,
) -> tuple[dict[str, list[str]], dict[str, str]]:
    """Resolve all @import: directives and return importable blocks + vars.

    Returns (imported_blocks, import_vars) where:
      - imported_blocks: {"AliasOrName": [action_lines...]}
      - import_vars: merged @var: from all imported files (lowest priority)

    Raises :class:`HuntImportError` when the import chain reaches
    ``_MAX_IMPORT_DEPTH`` levels (currently 10) to protect against
    pathological recursive imports.
    """
    if _depth >= _MAX_IMPORT_DEPTH:
        raise HuntImportError(
            f"Import depth limit ({_MAX_IMPORT_DEPTH}) reached while resolving "
            f"imports in {hunt_file}. Check for deep or indirect circular imports."
        )
    if seen_files is None:
        seen_files = set()
    seen_files.add(os.path.abspath(hunt_file))
    cwd = cwd or os.getcwd()

    imported_blocks: dict[str, list[str]] = {}
    import_vars: dict[str, str] = {}

    for directive in directives:
        try:
            source_path = resolve_source_path(directive.source, hunt_dir, cwd)
        except HuntImportError as e:
            raise HuntImportError(
                f"Cannot resolve '@import: {', '.join(directive.block_names)} "
                f"from {directive.source}' in {hunt_file}: {e}"
            ) from e

        exports, blocks, lib_vars = _extract_exported_blocks(
            source_path,
            seen_files=set(seen_files),
        )

        # Merge library vars (import-level, will be lowest priority)
        import_vars.update(lib_vars)

        # Determine which blocks to import
        if directive.block_names == ["*"]:
            # Wildcard: import all exported blocks
            if exports and exports != ["*"]:
                requested = exports
            else:
                # If @export: * or no exports declared at all with wildcard import
                if exports == ["*"] or not exports:
                    requested = list(blocks.keys())
                else:
                    requested = exports
        else:
            requested = directive.block_names

        for name in requested:
            if name not in blocks:
                available = ", ".join(blocks.keys()) or "(none)"
                raise HuntImportError(
                    f"Block '{name}' not found in '{directive.source}' "
                    f"(resolved: {source_path}). "
                    f"Available blocks: {available}"
                )

            # Check export access (if the file has explicit exports)
            if exports and exports != ["*"] and name not in exports:
                raise HuntImportError(
                    f"Block '{name}' exists in '{directive.source}' but is not exported. "
                    f"Exported blocks: {', '.join(exports)}"
                )

            # Apply alias if specified
            alias = directive.aliases.get(name, name)
            imported_blocks[alias] = blocks[name]

    return imported_blocks, import_vars


def validate_exports(hunt_path: str) -> list[str]:
    """Validate that all @export: names reference existing STEP blocks.

    Returns a list of warning messages for exports that have no matching block.
    """
    exports, blocks, _ = _extract_exported_blocks(hunt_path)
    warnings: list[str] = []
    if exports == ["*"]:
        return warnings
    for name in exports:
        if name not in blocks:
            warnings.append(f"@export: '{name}' has no matching STEP block in {hunt_path}")
    return warnings


def expand_use_directives(
    mission_lines: list[str],
    step_file_lines: list[int],
    imported_blocks: dict[str, list[str]],
) -> tuple[list[str], list[int]]:
    """Replace USE <BlockName> lines with imported block actions.

    Returns (expanded_lines, expanded_file_lines) where USE lines have been
    replaced by the corresponding imported actions.  File line numbers for
    expanded lines are set to 0 (synthetic/imported).
    """
    if len(mission_lines) != len(step_file_lines):
        raise HuntImportError(
            "Mission lines and step file line mappings are out of sync: "
            f"{len(mission_lines)} mission lines vs "
            f"{len(step_file_lines)} step file lines."
        )

    expanded: list[str] = []
    expanded_lines: list[int] = []

    _re_use = re.compile(r"^\s*(?:\d+\.\s*)?USE\s+(.+?)\s*$", re.IGNORECASE)

    for raw_line, file_line in zip(mission_lines, step_file_lines):
        m = _re_use.match(raw_line.strip())
        if m:
            block_name = m.group(1).strip()
            if block_name not in imported_blocks:
                raise HuntImportError(
                    f"USE '{block_name}' references an unknown import. "
                    f"Available imports: {', '.join(imported_blocks.keys()) or '(none)'}. "
                    f"Did you forget an @import: header?"
                )
            actions = imported_blocks[block_name]
            for action in actions:
                # Preserve 4-space indentation of expanded actions
                expanded.append(f"    {action}\n")
                expanded_lines.append(0)  # synthetic line
        else:
            expanded.append(raw_line)
            expanded_lines.append(file_line)

    return expanded, expanded_lines

File: core/pkg/dsl/imports.go
package dsl

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// ResolveImports loads and resolves all @import directives for a Hunt.
// It reads the imported .hunt files relative to the source file's directory,
// and merges imported STEP blocks (or all commands for wildcard imports)
// into the Hunt's available blueprints.
//
// Blueprint blocks are stored in hunt.Blueprints keyed by their STEP name.
// If aliases are specified, the alias name is used as the key.
//
// Circular imports are detected and reported as an error.
func ResolveImports(hunt *Hunt) error {
	visited := map[string]bool{}
	if hunt.SourcePath != "" {
		resolved, err := filepath.Abs(hunt.SourcePath)
		if err == nil {
			visited[resolved] = true
		}
	}
	return resolveImports(hunt, visited)
}

// resolveImports is the internal recursive implementation with cycle detection.
func resolveImports(hunt *Hunt, visited map[string]bool) error {
	if len(hunt.Imports) == 0 {
		return nil
	}

	if hunt.Blueprints == nil {
		hunt.Blueprints = make(map[string][]Command)
	}

	baseDir := filepath.Dir(hunt.SourcePath)
	if baseDir == "" {
		baseDir = "."
	}

	for _, imp := range hunt.Imports {
		importPath := filepath.Join(baseDir, imp.Source)

		// Resolve to absolute path for reliable cycle detection.
		absImportPath, err := filepath.Abs(importPath)
		if err != nil {
			absImportPath = importPath
		}

		if visited[absImportPath] {
			return fmt.Errorf("circular import detected: %s imports %s which creates a cycle",
				hunt.SourcePath, imp.Source)
		}

		if _, err := os.Stat(importPath); err != nil {
			return fmt.Errorf("import %q: file not found: %s", imp.Source, importPath)
		}

		imported, err := ParseFile(importPath)
		if err != nil {
			return fmt.Errorf("import %q: %w", imp.Source, err)
		}

		// Recurse with the visited set extended by the newly imported path.
		visitedChild := make(map[string]bool, len(visited)+1)
		for k, v := range visited {
			visitedChild[k] = v
		}
		visitedChild[absImportPath] = true
		if err := resolveImports(imported, visitedChild); err != nil {
			return fmt.Errorf("import %q: %w", imp.Source, err)
		}

		// Extract STEP blocks by their label.
		stepBlocks := extractStepBlocks(imported.Commands)

		if len(imp.Names) == 1 && imp.Names[0] == "*" {
			// Wildcard import: import all step blocks.
			for name, cmds := range stepBlocks {
				hunt.Blueprints[name] = cmds
			}
			// Also import file-level vars.
			for k, v := range imported.Vars {
				if _, exists := hunt.Vars[k]; !exists {
					hunt.Vars[k] = v
				}
			}
		} else {
			// Named imports: only import specified blocks.
			for _, name := range imp.Names {
				cmds, found := stepBlocks[name]
				if !found {
					// Try case-insensitive match.
					for k, v := range stepBlocks {
						if strings.EqualFold(k, name) {
							cmds = v
							found = true
							break
						}
					}
				}
				if !found {
					return fmt.Errorf("import %q: step block %q not found in %s",
						imp.Source, name, importPath)
				}

				key := name
				if alias, ok := imp.Aliases[name]; ok {
					key = alias
				}
				hunt.Blueprints[key] = cmds
			}
		}
	}
	return nil
}

// extractStepBlocks groups commands by their StepBlock label.
// It normalizes labels (e.g., "STEP 1: Login" -> "login") as keys.
func extractStepBlocks(cmds []Command) map[string][]Command {
	blocks := make(map[string][]Command)
	for _, cmd := range cmds {
		label := cmd.StepBlock
		if label == "" {
			continue
		}
		norm := normalizeStepName(label)
		blocks[norm] = append(blocks[norm], cmd)
	}
	return blocks
}

// normalizeStepName strips leading numbering and "STEP :" prefix, e.g.:
// "1. STEP 1: Login" -> "login"
// "STEP: Finalize" -> "finalize"
func normalizeStepName(s string) string {
	s = strings.ToUpper(s)
	// Strip "STEP"
	idx := strings.Index(s, "STEP")
	if idx >= 0 {
		s = s[idx+4:]
	}
	// Strip colon
	idx = strings.Index(s, ":")
	if idx >= 0 {
		s = s[idx+1:]
	}
	// Strip leading numbers/dots/whitespace
	s = strings.TrimLeft(s, " 0123456789.")
	return strings.ToLower(strings.TrimSpace(s))
}
